utils/zerotrie/src/builder/dense.rs

```diff
@@ -0,0 +1,177 @@
+            self.dense.push(Row {
+                prefix,
+                row_value_offset: min,
+                offsets,
+            });
+            for (suffix, value) in sparse_only.iter() {
+                self.primary.push((prefix, *suffix, *value));
+            }
+            Ok(())
+        } else {
+            // Create a sparse prefix
+            for (suffix, value) in values.iter() {
+                self.primary.push((prefix, *suffix, *value));
+            }
+            Ok(())
+        }
+    }
+
+    /// Assemble the intermediate structures into the final layout.
+    pub(crate) fn build(mut self) -> Result<ZeroAsciiDenseSparse2dTrieOwned, ZeroTrieBuildError> {
+        self.dense.sort();
+        let Ok(suffix_count) = DenseType::try_from(self.suffixes.len()) else {
+            return Err(ZeroTrieBuildError::CapacityExceeded);
+        };
+        let delimiter = self.delimiter as char;
+        let mut primary_contents = BTreeMap::new();
+        for (prefix, suffix, value) in self.primary.iter() {
+            if prefix.contains(delimiter) || suffix.contains(delimiter) {
+                debug_assert!(false, "handled earlier");
+                return Err(ZeroTrieBuildError::IllegalDelimiter);
+            }
+            let mut delimited = String::with_capacity(prefix.len() + suffix.len() + 1);
+            delimited.push_str(prefix);
+            delimited.push(delimiter);
+            delimited.push_str(suffix);
+            primary_contents.insert(delimited, *value);
+        }
+        let mut dense = Vec::<DenseType>::with_capacity(self.dense.len() * self.suffixes.len());
+        for (
+            row_index,
+            Row {
+                prefix,
+                row_value_offset,
+                offsets,
+            },
+        ) in self.dense.iter().enumerate()
+        {
+            primary_contents.insert(String::from(*prefix), row_index);
+            let mut prefix_with_delim = String::with_capacity(prefix.len() + 1);
+            prefix_with_delim.push_str(prefix);
+            prefix_with_delim.push(delimiter);
+            primary_contents.insert(prefix_with_delim, *row_value_offset);
+            dense.extend(offsets);
+        }
+        let suffixes = self
+            .suffixes
+            .iter()
+            .enumerate()
+            .map(|(column_index, suffix)| (*suffix, column_index))
+            .collect::<BTreeMap<&str, usize>>();
+        Ok(ZeroAsciiDenseSparse2dTrieOwned {
+            primary: ZeroTrieSimpleAscii::try_from_btree_map_str(&primary_contents)?,
+            suffixes: ZeroTrieSimpleAscii::try_from_btree_map_str(&suffixes)?,
+            dense: ZeroVec::from_slice_or_alloc(dense.as_slice()).into_owned(),
+            suffix_count,
+            delimiter: self.delimiter,
+        })
+    }
+}
+
+impl ZeroAsciiDenseSparse2dTrieOwned {
+    /// Builds one of these from a two-dimensional BTreeMap and a delimiter.
+    pub fn try_from_btree_map_str(
```

Comment:
nit: please document the soft invariant that the inner btree values should be all in the same locality for maximum performance

Reply:
I added docs about the locality invariant. I called it "in a compact range".